src/XML/src/FMI3/fmi3_xml_terminals_and_icons.c

```diff
@@ -0,0 +1,238 @@
+            fmi3_xml_parse_fatal(context, "fmiTerminalsAndIcons must be the root XML element");
+            return -1;
+        }
+        jm_log_verbose(context->callbacks, module, "Parsing XML element fmiTerminalsAndIcons");
+
+        ret = fmi3_xml_parse_attr_as_string(context, fmi3_xml_elmID_fmiTerminalsAndIcons, fmi_attr_id_fmiVersion, 1 /* required */,
+                                            &(termIcon->fmi3_xml_standard_version));
+        if (ret) {
+            return ret;
+        }
+        // Error check: fmiVersion in terminalsAndIcons and modelDescription must match
+        const char* modelDescrfmiVersion = fmi3_xml_get_model_standard_version(termIcon->md);
+        const char* termIconfmiVersion = jm_vector_char2string(&(termIcon->fmi3_xml_standard_version));
+        if (strcmp(modelDescrfmiVersion, termIconfmiVersion)) {
+            fmi3_xml_parse_fatal(context,
+                                 "Mismatch of attribute 'fmiVersion' in modelDescription.xml: '%s' and terminalsAndIcons.xml: '%s'.", 
+                                 modelDescrfmiVersion,
+                                 termIconfmiVersion);
+            return -1;
+        }
+
+        return ret;
+    } else { // post process <fmiTerminalsAndIcons>
+        ;
+    }
+    return 0;
+}
+
+int fmi3_xml_handle_Terminals(fmi3_xml_parser_context_t* context, const char* data) {
+    fmi3_xml_terminals_and_icons_t* termIcon = context->termIcon;
+    if (!data) {
+        ;
+    } else { // post process <Terminals>
+        // Create terminalsByName
+        size_t nVars = jm_vector_get_size(jm_voidp)(&termIcon->terminalsOrigOrder);
+
+        jm_vector_resize(jm_named_ptr)(&termIcon->terminalsByName, nVars);  // Allocate a lot of memory to avoid resizing
+        jm_vector_resize(jm_named_ptr)(&termIcon->terminalsByName, 0);      // Make sure we start pushback at index=0
+        for (size_t i = 0; i < nVars; i++) {
+            jm_named_ptr named;
+            fmi3_xml_terminal_t* term = jm_vector_get_item(jm_voidp)(&termIcon->terminalsOrigOrder, i);
+            named.name = term->name;
+            named.ptr = term;
+            jm_vector_push_back(jm_named_ptr)(&termIcon->terminalsByName, named);
+        }
+        jm_vector_qsort(jm_named_ptr)(&termIcon->terminalsByName, jm_compare_named);
+
+        // Error check, Terminal names are must be unique
+        if (nVars > 0) {  // nVars=0 would cause integer overflow in the loop condition
+            for (size_t i = 0; i < nVars-1; ++i) {
+                const char* name1 = jm_vector_get_item(jm_named_ptr)(&termIcon->terminalsByName, i).name;
+                const char* name2 = jm_vector_get_item(jm_named_ptr)(&termIcon->terminalsByName, i+1).name;
+                if(strcmp(name1, name2) == 0) {
+                    fmi3_xml_parse_fatal(context, 
+                            "Two terminals with the same name '%s' found. This is not allowed.", name1);
```

Comment:
```suggestion
                            "Two terminals with the same name '%s' found. This is not allowed by the specification.", name1);
```
Consider this optional.